### PR TITLE
Add support for multiple data types (will break up into smaller pull requests)

### PR DESCRIPTION
**Pull request by hcho3**
Addresses #95 and #111.

Trying again, since #130 failed. This time, I made the `Model` class to be polymorphic. This way, the amount of pointer indirection is minimized.

Summary: `Model` is an opaque container that wraps the polymorphic handle `ModelImpl<ThresholdType, LeafOutputType>`. The handle in turn stores the list of trees `Tree<ThresholdType, LeafOutputType>`. To unbox the `Model` container and obtain `ModelImpl<ThresholdType, LeafOutputType>`, use `Model::Dispatch(<lambda expression>)`. 

Also, upgrade to C++14 to access the generic lambda feature, which proved to be very useful in the dispatching logic for the polymorphic `Model` class.

EDIT. I will break up this PR into smaller PRs, once I get the whole system working together correctly.

TODOs
- [x] Turn the Model and Tree classes into template classes
- [x] Revise the string templates so that correct data types are used in the generated C code
- [x] Rewrite the model builder class
- [x] Revise the zero-copy serializer
- [x] Create an abstract matrix class that supports multiple data types (float32, float64 for now).
- [x] Redesign the C runtime API, using the abstract matrix class.
- [x] Ensure accuracy of scikit-learn models. To achieve the best results, use float32 for the input matrix and float64 for the split thresholds and leaf outputs.
- [x] Revise the JVM runtime.

### CODE REVIEW

**Comment by hcho3** [comment]
#198, #199, #201, and #203 have been merged into a separate branch `multi_type_support_refactor`. Once the XGBoost JSON parser (#202) is merged into the mainline, the refactor branch will be also merged into the mainline.
